proposals/0083-relax-entry-constraints.md

```diff
@@ -0,0 +1,62 @@
+---
+simd: '0083'
+title: Relax Entry Constraints
+authors:
+  - Andrew Fitzgerald (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-11-02
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Remove the constraint that transaction entries must not conflict with each
+other.
+
+## Motivation
+
+The current constraint that transaction entries must not conflict with each
+other is unnecessary.
+It is a constraint, that made the current labs-client implementation easier,
+but is not required for the protocol to function correctly.
+The removal of this constraint simplifies the protocol, and allows more
+flexibility in the implementation of both block-produciton and
+block-validation.
+
+## Alternatives Considered
+
+1. Do nothing
+    - This is the simplest option, as we could leave the protocol as is.
+    However, this leaves the protocol more complex than it needs to be.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Currently, if a transaction entry within a block contains transactions that
+conflict, the entire block is marked invalid.
+The proposal is that this constraint is removed entirely, and entries are
+allowed to contain transactions that conflict with each other.
+
```

Comment:
We implemented and tested the required changes to enable self conflicting entries in our lab setup and this improves the performance a lot depending on how you schedule the transactions on every execution thread.
The changes include:

- changing account locking function
- carrying temporary account state within the execution with every transaction in an entry

As the functions changed are used by both the executor and validator, so it doesn't break any consensus and rules of the chain, and is backward compatible.
These changes can be implemented through a feature gate.

@apfitzge we can add these points to the detailed design if you think it's appropriate.

Reply:
> As the functions changed are used by both the executor and validator, so it doesn't break any consensus and rules of the chain, and is backward compatible.
These changes can be implemented through a feature gate.

This changes the structure of blocks and that directly affects consensus. In Firedancer today you would break validation criteria for a block and there would be a hard fork from Solana Labs.

Reply:
@lheeger-jump is correct, this is a consensus breaking change. If you produce a block that contains entries with conflicting transactions, existing validators will mark it as invalid. For this reason, it needs to sit behind a feature, and only be activated when both validator clients have implemented the feature-gate **and** those versions have been rolled out to cluster(s).

> @apfitzge we can add these points to the detailed design if you think it's appropriate.

This doesn't seem to be appropriate for a SIMD. SIMDs should be independent of any specific client's implementation. The issue in the labs repo you created would be a more appropriate place to share implementation-specific code and any results.